### PR TITLE
Simplify, avoid deprecated API MODLOGSAML-122

### PR DESCRIPTION
**Pull request by adamdickmeiss**
Approx 250 lines removed from src/main.

Increases total coverage from 75.1% to 81.4%

https://issues.folio.org/browse/MODLOGSAML-122

### CODE REVIEW

File: src/main/java/org/folio/config/ConfigurationsClient.java [hunk 1]
```diff
@@ -139,96 +88,50 @@ protected static void verifyOkapiHeaders(OkapiHeaders okapiHeaders) throws Missi
...
+              throw new RuntimeException("Could not store entry code=" + code + " value=" + value + " status= " + res.statusCode());
+            }
+            return null;
+          });
+      });
   }
 
+  public static Future<JsonArray> checkConfig(Vertx vertx, OkapiHeaders okapiHeaders, String query) {
+    verifyOkapiHeaders(okapiHeaders);
+    String encodedQuery = URLEncoder.encode(query, StandardCharsets.UTF_8);
```
Comment (julianladisch) [inline src/main/java/org/folio/config/ConfigurationsClient.java:114]:
```suggestion
    CharSequence encodedQuery = PercentCodec.encode(query);
```
https://github.com/folio-org/raml-module-builder/blob/v33.1.1/util/src/main/java/org/folio/util/PercentCodec.java#L91-L108
URLEncoder encodes HTML forms, not URLs: https://docs.oracle.com/javase/10/docs/api/java/net/URLEncoder.html

File: src/main/java/org/folio/config/ConfigurationsClient.java [hunk 2]
```diff
@@ -139,96 +88,50 @@ protected static void verifyOkapiHeaders(OkapiHeaders okapiHeaders) throws Missi
...
+          .putHeader(XOkapiHeaders.TOKEN, okapiHeaders.getToken())
+          .putHeader(XOkapiHeaders.URL, okapiHeaders.getUrl())
+          .putHeader(XOkapiHeaders.TENANT, okapiHeaders.getTenant())
+          .sendJsonObject(requestBody)
+          .map(res -> {
+            if (res.statusCode() != 200 && res.statusCode() != 201 && res.statusCode() != 204) {
+              throw new RuntimeException("Could not store entry code=" + code + " value=" + value + " status= " + res.statusCode());
+            }
+            return null;
+          });
```
Comment (julianladisch) [inline src/main/java/org/folio/config/ConfigurationsClient.java:108]:
```suggestion
          .expect(ResponsePredicate.status(201, 204))
          .sendJsonObject(requestBody)
          .mapEmpty();
```
https://vertx.io/docs/vertx-web-client/java/#_predefined_predicates
Neither PUT nor POST return 200 on success.

File: src/main/java/org/folio/config/SamlClientLoader.java [hunk 3]
```diff
@@ -189,39 +190,41 @@ private SamlClientLoader() {
   }
 
 
-  private static SAML2Client configureSaml2Client(String okapiUrl, String tenantId, String idpUrl, String keystorePassword, String actualPrivateKeyPassword, String keystoreFileName, String samlBinding) {
+  private static SAML2Client configureSaml2Client(String okapiUrl, String tenantId, String idpUrl, String keystorePassword, String actualPrivateKeyPassword,
+    String keystoreFileName, String samlBinding, Context vertxContext) {
```
Comment (julianladisch) [inline src/main/java/org/folio/config/SamlClientLoader.java:194]:
```suggestion
      String keystoreFileName, String samlBinding, Context vertxContext) {

```

File: src/main/java/org/folio/rest/impl/SamlAPI.java [hunk 4]
```diff
@@ -185,259 +189,192 @@ public void postSamlCallback(RoutingContext routingContext, Map<String, String>
...
+        // Get user id
+        List<?> samlAttributeList = (List<?>) credentials.getUserProfile().getAttribute(samlAttributeName);
+        if (samlAttributeList == null || samlAttributeList.isEmpty()) {
+          throw new UserErrorException("SAML attribute doesn't exist: " + samlAttributeName);
+        }
+        final String samlAttributeValue = samlAttributeList.get(0).toString();
+
+        final String usersCql = userPropertyName +
+          "=="
+          + QUOTATION_MARK_CHARACTER + samlAttributeValue + QUOTATION_MARK_CHARACTER;
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/SamlAPI.java:210]:
```suggestion
        final String usersCql = userPropertyName + "==" + StringUtil.cqlEncode(samlAttributeValue);
```
In addition userPropertyName should be checked against a white list.

File: src/main/java/org/folio/rest/impl/SamlAPI.java [hunk 5]
```diff
@@ -185,259 +189,192 @@ public void postSamlCallback(RoutingContext routingContext, Map<String, String>
...
-              }
-            });
-        }
+    Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+
+    regenerateSaml2Config(routingContext, vertxContext)
+      .compose(metadata ->
+        ConfigurationsClient.storeEntry(vertxContext.owner(), OkapiHelper.okapiHeaders(okapiHeaders), SamlConfiguration.METADATA_INVALIDATED_CODE, "false")
+          .compose(configurationEntryStoredEvent ->
+            Base64Util.encode(vertxContext, metadata)
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/SamlAPI.java:297]: In Base64Util a method `public static String encode(String input)` should be created that avoids the Future, the executeBlocking and the Buffer.

File: src/main/java/org/folio/config/ConfigurationsClient.java (hunk 1 above)
Comment (adamdickmeiss) [reply src/main/java/org/folio/config/ConfigurationsClient.java:114]: ok

File: src/main/java/org/folio/config/ConfigurationsClient.java (hunk 2 above)
Comment (adamdickmeiss) [reply src/main/java/org/folio/config/ConfigurationsClient.java:108]: OK. It's actually 201, 205. Weird..

File: src/main/java/org/folio/rest/impl/SamlAPI.java [hunk 6]
```diff
@@ -185,259 +189,194 @@ public void postSamlCallback(RoutingContext routingContext, Map<String, String>
...
-            });
-        }
-      });
+  private Future<SamlConfig> storeConfigEntries(RoutingContext rc, OkapiHeaders parsedHeaders, Map<String, String> updateEntries,
+    Context vertxContext) {
+    return ConfigurationsClient.storeEntries(vertxContext.owner(), parsedHeaders, updateEntries)
+      .compose(configurationSavedEvent ->
+        findSaml2Client(rc, true, true, vertxContext)
+          .map(configurationLoadEvent -> configToDto(configurationLoadEvent.getConfiguration())
+          ));
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/SamlAPI.java:379]:
```suggestion
      .compose(x ->findSaml2Client(rc, true, true, vertxContext))
      .map(configurationLoadEvent -> configToDto(configurationLoadEvent.getConfiguration()));
```

File: src/main/java/org/folio/util/Base64Util.java [hunk 7]
```diff
@@ -24,13 +24,10 @@ private Base64Util() {
    * @return Buffer bytes of Base64 string
    */
   public static Future<Buffer> encode(Context context, String content) {
-
-    Promise<Buffer> promise = Promise.promise();
-    context.<Buffer>executeBlocking(blockingCode -> {
+    return context.executeBlocking(blockingCode -> {
```
Comment (julianladisch) [inline src/main/java/org/folio/util/Base64Util.java:27]: Base64 encoding is a very fast operation, no need to wrap into executeBlocking.

File: src/main/java/org/folio/util/UrlUtil.java [hunk 8]
```diff
@@ -18,27 +17,27 @@ private UrlUtil() {
...
-        return UrlCheckResult.failResult("ConnectException: " + cause.getMessage());
-      }
-      return UrlCheckResult.failResult("Unexpected error: " + cause.getMessage());
-    });
+      .map(httpResponse -> {
+        String contentType = httpResponse.getHeader("Content-Type");
+        if (contentType == null || ! contentType.contains("xml")) {
+          throw new RuntimeException("Response content-type is not XML");
+        }
+        return null;
```
Comment (julianladisch) [inline src/main/java/org/folio/util/UrlUtil.java:32]:
```suggestion
        return (Void) null;
```
avoids mapEmpty()

File: src/test/java/org/folio/rest/impl/SamlAPITest.java [hunk 9]
```diff
@@ -416,7 +430,7 @@ public void callbackEndpointTests() throws IOException {
       .post("/saml/callback")
       .then()
       .statusCode(302)
-      .header("Location", containsString(URLEncoder.encode(testPath, "UTF-8")))
+      .header("Location", containsString(URLEncoder.encode(testPath, StandardCharsets.UTF_8)))
```
Comment (julianladisch) [inline src/test/java/org/folio/rest/impl/SamlAPITest.java:433]:
```suggestion
      .header("Location", containsString(PercentCodec.encode(testPath)))
```

File: src/test/java/org/folio/rest/impl/SamlAPITest.java [hunk 10]
```diff
@@ -453,7 +469,64 @@ public void callbackEndpointTests() throws IOException {
...
+      .header(TENANT_HEADER)
+      .header(TOKEN_HEADER)
+      .header(OKAPI_URL_HEADER)
+      .cookie(SamlAPI.RELAY_STATE, cookie)
+      .formParam("SAMLResponse", "saml-response")
+      .formParam("RelayState", relayState)
+      .post("/saml/callback")
+      .then()
+      .statusCode(302)
+      .header("Location", containsString(URLEncoder.encode(testPath, StandardCharsets.UTF_8)))
```
Comment (julianladisch) [inline src/test/java/org/folio/rest/impl/SamlAPITest.java:526]:
```suggestion
      .header("Location", containsString(PercentCodec.encode(testPath)))
```

File: src/test/java/org/folio/util/IdpMock.java [hunk 11]
```diff
@@ -22,14 +25,8 @@ public void start(Promise<Void> promise) {
...
-    System.out.println("Running IdpMock on port " + port);
-    server.requestHandler(router).listen(port, result -> {
-      if (result.failed()) {
-        promise.fail(result.cause());
-      } else {
-        promise.complete();
-      }
-    });
+    log.info("Running IdpMock on port {}", port);
+    server.requestHandler(router).listen(port).onComplete(x -> promise.handle(x.mapEmpty()));
```
Comment (julianladisch) [inline src/test/java/org/folio/util/IdpMock.java:29]:
```suggestion
    server.requestHandler(router).listen(port).<Void>mapEmpty().onComplete(promise);
```

File: src/test/java/org/folio/util/MockJson.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  private void handle(RoutingContext context) {
+    HttpServerRequest request = context.request();
+    HttpServerResponse response = context.response();
+    String method = request.method().name();
+    String uri = request.uri();
+    for (int i = 0; i < mocks.size(); i++) {
+      JsonObject entry = mocks.getJsonObject(i);
+      if (method.compareToIgnoreCase(entry.getString("method", "get")) == 0
```
Comment (julianladisch) [inline src/test/java/org/folio/util/MockJson.java:37]:
```suggestion
      if (method.equalsIgnoreCase(entry.getString("method", "get"))
```

File: src/test/java/org/folio/util/MockJson.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  public void start(Promise<Void> promise) {
+    final int port = context.config().getInteger("http.port");
+
+    log.info("Running Mock JSON on port {}", port);
+
+    Router router = Router.router(vertx);
+    router.routeWithRegex("/.*").handler(this::handle);
+    vertx.createHttpServer().requestHandler(router).listen(port).onComplete(x -> promise.handle(x.mapEmpty()));
```
Comment (julianladisch) [inline src/test/java/org/folio/util/MockJson.java:70]:
```suggestion
    vertx.createHttpServer().requestHandler(router).listen(port).<Void>mapEmpty().onComplete(promise);
```

File: src/test/java/org/folio/util/UrlUtilTest.java [hunk 14]
```diff
@@ -52,34 +52,32 @@ public static void afterOnce(TestContext context) {
...
-      }));
+      .onComplete(context.asyncAssertFailure(cause ->
+        // check locale independent prefix only.
+        assertThat(cause.getMessage(), startsWith("ConnectException: "))
+      ));
   }
 
   @Test
-  public void checkIdpUrlUnexpectedError(TestContext context) {
+  public void checkIdpUrlNoCnotentType(TestContext context) {
```
Comment (julianladisch) [inline src/test/java/org/folio/util/UrlUtilTest.java:69]:
```suggestion
  public void checkIdpUrlNoContentType(TestContext context) {
```

File: src/main/java/org/folio/rest/impl/SamlAPI.java (hunk 5 above)
Comment (adamdickmeiss) [reply src/main/java/org/folio/rest/impl/SamlAPI.java:297]: sure

File: src/test/java/org/folio/rest/impl/SamlAPITest.java (hunk 9 above)
Comment (adamdickmeiss) [reply src/test/java/org/folio/rest/impl/SamlAPITest.java:433]: Using PercentCodec.encodeAsString instead.

File: src/main/java/org/folio/rest/impl/SamlAPI.java (hunk 4 above)
Comment (adamdickmeiss) [reply src/main/java/org/folio/rest/impl/SamlAPI.java:210]: I'll do these things in #90.

**Review by julianladisch — APPROVED** [review]
Looks good, thanks!

**Pull request merged**